source/common/upstream/cds_api_impl.cc

```diff
@@ -66,7 +66,34 @@ void CdsApiImpl::onConfigUpdate(const ResourceVector& resources, const std::stri
     }
   }
 
-  version_info_ = version_info;
+  whole_update_version_info_ = version_info;
+  runInitializeCallbackIfAny();
+}
+
+void CdsApiImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+    const std::string& system_version_info) {
+  cm_.adsMux().pause(Config::TypeUrl::get().ClusterLoadAssignment);
+  Cleanup eds_resume([this] { cm_.adsMux().resume(Config::TypeUrl::get().ClusterLoadAssignment); });
+
+  for (const auto& resource : added_resources) {
+    envoy::api::v2::Cluster cluster =
+        MessageUtil::anyConvert<envoy::api::v2::Cluster>(resource.resource());
+    MessageUtil::validate(cluster);
+    if (cm_.addOrUpdateCluster(cluster, resource.version())) {
+      ENVOY_LOG(debug, "cds: add/update cluster '{}'", cluster.name());
+    }
+    // TODO(fredlas) catch exceptions from this block, and add the offending resource to a "resource
+    // updates that had problems" pile, to go into a larger partial rejection exception.
+  }
+  for (auto resource_name : removed_resources) {
+    if (cm_.removeCluster(resource_name)) {
+      ENVOY_LOG(debug, "cds: remove cluster '{}'", resource_name);
+    }
+  }
+
+  whole_update_version_info_ = system_version_info;
```

Comment:
I continue to be concerned about the duplication between the incremental and regular xDS `onConfigreUpdate`. I know you have a `TODO` in the interface definition, but, for example, we merged changes to regular `onConfigUpdate` that require reflection here last night.

Do you think you can unify `onConfigUpdate` in this current PR?

Reply:
I think they can only be unified at the time when incremental is fully implemented for all xDS types, so that vanilla xDS basically doesn't exist anymore. If you still want this all to start with this PR being just a working implementation of CDS, I think it has to be this way.

Actually, if we want the code to implement vanilla by just translating to incremental, doesn't that completely drop support for xDS servers that don't speak incremental? Since Envoy would no longer be willing to deal in (non-Delta)Discovery{Request,Response}.

Reply:
I'm talking here just about the CDS duplication. Can't we factor the logic out between incremental and non-incremental implementations?

Reply:
Ah, I see, not complete unification, but rather sharing the logic that actually modifies the ClusterManager, so that the implementations don't drift out of sync? Done.